Zend/zend_call_stack.c

```diff
@@ -432,6 +435,28 @@ static bool zend_call_stack_get_macos(zend_call_stack *stack)
 }
 #endif /* defined(__APPLE__) && defined(HAVE_PTHREAD_GET_STACKADDR_NP) */
 
+#if defined(HAVE_PTHREAD_STACKSEG_NP)
+static bool zend_call_stack_get_openbsd(zend_call_stack *stack)
+{
+	stack_t ss;
+
+	if (pthread_stackseg_np(pthread_self(), &ss) != 0) {
+		return false;
+	}
+
+	stack->base = (char *)ss.ss_sp - ss.ss_size;
+	// TODO may need more leeway for the main thread
+	stack->max_size = ss.ss_size;
```

Comment:
Are we able to get the actual `ss_sp` for the main thread? Does the stack reside in its own memory mapping, and can we find the address of this mapping?

Otherwise, do we have a way to estimate the right size for the leeway? Does it depend on the arguments and environment size?

Reply:
> Are we able to get the actual `ss_sp` for the main thread? 

Yes, if pthread_stackseg_np succeeds you always get it.

> Does the stack reside in its own memory mapping, and can we find the address of this mapping?

> Otherwise, do we have a way to estimate the right size for the leeway? Does it depend on the arguments and environment size?

Looking at openbsd source, it seems [the proper calculation is done already since years now](https://github.com/openbsd/src/commit/7e9a0975a41eeeba58810cd67dabb4af33d47c0c#diff-776b65a5fc714a527c9ed1102d00c5d732dac696a6b0b16a7c35c5de67fcfe50R105), back in the days you needed the alignment yourself. What you re looking at is the main thread case and it depends on the arguments and environment indeed, done in the [kernel side of things here](https://github.com/openbsd/src/blob/dcb03dac24b6a52fa9b433eceeaac7ff6f5182e3/lib/libkvm/kvm_proc.c#L258).

Reply:
Is it possible that the alignment is not enough to take `sgap` and `sizeof(arginfo)` into account? `VM_PSSTRINGS` is [`vm->vm_minsaddr - sizeof(arginfo) - sgap`](https://github.com/openbsd/src/blob/master/sys/kern/kern_exec.c#L485), and sgap appears to be in [0..pagesize], so I understand that `VM_PSSTRINGS` can be more than 1 page way from `vm_minsaddr`. In that case, we would need to add 1 page of leeway.